mixnet/nomssip.py

```diff
@@ -0,0 +1,125 @@
+import asyncio
+import hashlib
+from dataclasses import dataclass
+from enum import Enum
+from typing import Awaitable, Callable
+
+from mixnet.connection import DuplexConnection, MixSimplexConnection, SimplexConnection
+
+
+class Nomssip:
+    """
+    A NomMix gossip channel that broadcasts messages to all connected peers.
+    Peers are connected via DuplexConnection.
+    """
+
+    @dataclass
+    class Config:
+        transmission_rate_per_sec: int
+        peering_degree: int
+        msg_size: int
+
+    config: Config
+    conns: list[DuplexConnection]
+    # A handler to process inbound messages.
+    handler: Callable[[bytes], Awaitable[None]]
+    # A set of packet hashes to prevent gossiping/processing the same packet twice.
+    packet_cache: set[bytes]
+
+    def __init__(
+        self,
+        config: Config,
+        handler: Callable[[bytes], Awaitable[None]],
+    ):
+        self.config = config
+        self.conns = []
+        self.handler = handler
+        self.packet_cache = set()
+        # A set just for gathering a reference of tasks to prevent them from being garbage collected.
+        # https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task
+        self.tasks = set()
+
+    def add_conn(self, inbound: SimplexConnection, outbound: SimplexConnection):
+        if len(self.conns) >= self.config.peering_degree:
+            # For simplicity of the spec, reject the connection if the peering degree is reached.
+            raise ValueError("The peering degree is reached.")
+
+        noise_packet = self.__build_packet(
+            self.PacketType.NOISE, bytes(self.config.msg_size)
```

Comment:
do we send empty bytes? Wouldn't this be a pattern that could be use to denoise?

Reply:
We are assuming TLS on the connection level so for the observer it will be indistinguishable from real traffic. Peers must know that we are sending noise to let them drop that, so we can use zeros here or any known random sequence.

Reply:
Yes. Even if we use random sequence (instead of zeros), packets will be distinguishable if we don't have secure transport such as TLS. So here, we assume that we will always have secure transport.